File: go/backtracking/1593_max_unique_split.go
package backtracking

// maxUniqueSplit 拆分字符串使唯一子字符串的数目最大
func maxUniqueSplit(s string) int {
	// 回溯

	// 子串最终数目
	var res = 0
	// 搜索过程中的字符串子集，不能重复要用集合
	path := make(map[string]bool)

	// backtrack 回溯搜索各种子集的可能
	// idx 搜索开始的下标, count 子串的个数
	var backtrack func(idx, count int)
	backtrack = func(idx, count int) {

		// 剪枝, 剩余的字符个数小于 len(s)- idx
		// + 已经拆分的子集 count
		// 小于当前res 则没有必要继续递归搜索下去了
		if len(s)-idx+count <= res {
			return
		}

		if idx == len(s) {
			res = max(res, count)
			return
		}

		for i := idx; i < len(s); i++ {
			subStr := s[idx : i+1]
			if !path[subStr] {
				path[subStr] = true
				backtrack(i+1, count+1)
				path[subStr] = false
			}
		}
	}

	backtrack(0, 0)
	return res
}

File: rust/src/backtracking/_1593_max_unique_split.rs
use std::collections::HashSet;

/// 拆分字符串使唯一子字符串的数目最大
pub fn max_unique_split(s: String) -> i32 {
    if s.is_empty() {
        return 0;
    }

    // 子串最终的数目
    let mut res = 0;

    // 搜索过程中的字符串子集，不能重复要用集合
    let mut path = HashSet::new();
    let s: Vec<char> = s.chars().collect();
    backtrack(&s, 0, 0, &mut path, &mut res);
    res
}

fn backtrack(s: &[char], idx: usize, count: i32, path: &mut HashSet<Vec<char>>, res: &mut i32) {
    // 剪枝
    if s.len() - idx + count as usize <= *res as usize {
        return;
    }

    if idx >= s.len() {
        *res = (*res).max(count);
        return;
    }

    for i in idx..s.len() {
        let sub_str = &s[idx..i + 1].to_vec();
        if !path.contains(sub_str) {
            path.insert(sub_str.clone());
            backtrack(s, i + 1, count + 1, path, res);
            path.remove(sub_str);
        }
    }
}
